Modules/itertoolsmodule.c

```diff
@@ -193,47 +193,42 @@ batched_traverse(batchedobject *bo, visitproc visit, void *arg)
 static PyObject *
 batched_next(batchedobject *bo)
 {
-    Py_ssize_t i;
-    Py_ssize_t n = bo->batch_size;
```

Comment:
I don't get why any of this code needs to change.  It was short, clear, and highly optimized.  Unless something is broken, please leave it as is.

Reply:
Note, the normal path here is that we know the *exact* size of the result tuple.  Using overallocations and resizes just makes this code worse.

Reply:
> I don't get why any of this code needs to change. It was short, clear, and highly optimized. Unless something is broken, please leave it as is.

The bug being discussed is the fact that tuple which is being populated is exposed to the GC. See related issue for the rationale. This API is a draft, and deciding to fix the issue or not is being discussed.

Reply:
_PyTupleBuilder_Init(n) does more or less call PyTuple_New(n) as the current implementation. _PyTupleBuilder_Finish() only calls _PyTuple_Resize() if the iterator yielded les than n items.

For n <= 16, the proposed implementation uses a small buffer allocated on the stack , rather than calling PyTuple_New(n), and only allocate a tuple in _PyTupleBuilder_Finish().

These _PyTupleBuilder functions are implemented as static inline functions.

I agree that performance matters and a benchmark can be run to measure the real performance. But this PR is more to see how the API would look like. If batched_traverse() is changed, I would prefer to do it in a separated PR.